### PR TITLE
Update the p256 crate

### LINKED ISSUE #561: Upgrade p256 to >= 0.13
p256 `v0.12.0` is ancient: more than 2 years old!

https://github.com/tkhq/qos/blob/5f93df779d06853c2240c3fbb43d3e0d72fd8c4f/src/qos_p256/Cargo.toml#L14

Image

Recent crypto projects won't accept p256 <0.13...but more generally: time to upgrade.

### PR DESCRIPTION
**Pull request by c410-f3r**
## Summary & Motivation (Problem vs. Solution)

Fix #561

Also updates `yubikey`, `p384` and `x509-cert`.

## Pre merge check list

- [ ] Update CHANGELOG.MD

### CODE REVIEW

File: src/qos_p256/src/encrypt.rs [hunk 1]
```diff
@@ -100,15 +100,15 @@ impl P256EncryptPair {
 	/// Deserialize key from raw scalar byte slice.
 	pub fn from_bytes(bytes: &[u8]) -> Result<Self, P256Error> {
 		Ok(Self {
-			private: SecretKey::from_be_bytes(bytes)
```
Comment (c410-f3r) [inline src/qos_p256/src/encrypt.rs:103]: https://github.com/RustCrypto/traits/pull/1220

> The methods use a default big endian serialization. Now that curves can customize the endianness used for encoding field elements, this eliminates the need to have separate from_be_bytes/from_le_bytes and to_be_bytes/to_le_bytes methods. Instead the endianness can be known a priori, which eliminates the need for the caller to select a specific endianness. As such, this PR folds these methods (defined on e.g. ScalarPrimitive and NonZeroScalar) into from_bytes and to_bytes.

According to the PR, `from_slice` is a drop-in replacement for `from_be_bytes`. Same thing for `to_be_bytes`.

**Comment by c410-f3r** [comment]
Hello @cr-tk 

Well, I wish I could just update `p256`. Maybe I should have created an explanation when this PR became a draft.

My initial local development didn't catch https://github.com/tkhq/qos/actions/runs/16497458041/job/46646614643 and such a thing made me think just updating `p256`  would be enough. Unfortunately, `yubikey` is tightly coupled to specific versions of `x509-cert`, `p384` and `p256`. In other words, stuff won't compile.

```bash
    Updating crates.io index
error: failed to select a version for `signature`.
    ... required by package `x509-cert v0.2.3`
    ... which satisfies dependency `x509-cert = "^0.2.3"` of package `yubikey v0.8.0`
    ... which satisfies dependency `yubikey = "^0.8"` of package `qos_client v0.1.0
    ... which satisfies path dependency `qos_client` (locked to 0.1.0) of package `integration v0.1.0`
versions that meet the requirements `^2.1.0` are: 2.2.0, 2.1.0

all possible versions conflict with previously selected packages.

  previously selected package `signature v2.0.0`
    ... which satisfies dependency `signature = "^2"` of package `yubikey v0.8.0`
    ... which satisfies dependency `yubikey = "^0.8"` of package `qos_client v0.1.0
    ... which satisfies path dependency `qos_client` (locked to 0.1.0) of package `integration v0.1.0

failed to select a version for `signature` which could resolve this conflict
```

> One of your intermediary commits (now overwritten) changed the targeted Rust version, which is currently not what we're planning to do - towards nightly or otherwise.

A misunderstanding that luckily was identified and reverted.

> Additionally, please provide context on changes to cryptography code such as the certificate handling and pubkey handling, like you did on https://github.com/tkhq/qos/pull/564/files#r2228444460 already.

Every crypto primitive will be historically reviewed to maintain the same set of functionalities. Unfortunately, this takes some time and today I am no longer available. In two or three days everything shall be ready for a review. Sound good?

File: src/qos_client/src/yubikey.rs [hunk 2]
```diff
@@ -97,18 +106,15 @@ pub fn generate_signed_certificate(
...
 	yubikey.verify_pin(pin).map_err(YubiKeyError::FailedToVerifyPin)?;
-	Certificate::generate_self_signed(
+	Certificate::generate_self_signed::<_, p256::NistP256>(
 		yubikey,
 		slot,
-		serial,
-		None, // not_after is none so this never expires
-		&[RelativeDistinguishedName::organization("Turnkey")],
+		SerialNumber::new(&serial)?,
+		Validity::from_now(Duration::from_secs(u64::MAX))?, // not_after is none so this never expires
```
Comment (cr-tk) [inline src/qos_client/src/yubikey.rs]: With this change, the inline code comment is no longer accurate.

File: src/qos_client/src/yubikey.rs [hunk 3]
```diff
@@ -158,18 +165,15 @@ pub fn import_key_and_generate_signed_certificate(
...
```
Comment (cr-tk) [inline src/qos_client/src/yubikey.rs]: With this change, the inline code comment is no longer accurate.

File: src/qos_p256/src/encrypt.rs (hunk 1 above)
Comment (cr-tk) [reply src/qos_p256/src/encrypt.rs:103]: I think this is correct. Notably, the quoted description does not mention `::from_slice()`, only `::to_bytes()`, but `::from_slice()` uses `::to_bytes()` internally. So I think this still holds.

**Comment by c410-f3r** [comment]
> @c410-f3r Thanks for the feedback and explanation. Updating `p256` is indeed a bit of a headache.
> 
> We propose a two-stage approach:
> 
>     1. Our PR [Selectively update dependencies, documentation, rework code #565](https://github.com/tkhq/qos/pull/565) lays the groundwork for the `p256` update, including changes to `x509-cert` and `p384` which can be changed in isolation
> 
>     2. This PR [Update the p256 crate #564](https://github.com/tkhq/qos/pull/564) changes `p256`, `yubikey`, and a minimal set of necessary code positions on top of (1)
> 
> 
> This way, we reduce the complexity + review complexity of this PR, for example by removing changes to `qos_nsm` from the scope, and by bringing in a number of the necessary crate dependencies earlier.

Sure, sounds like a good plan. I will rebase this PR when #565 lands.

File: src/qos_client/src/yubikey.rs [hunk 4]
```diff
@@ -65,6 +67,14 @@ pub enum YubiKeyError {
...
 	FailedToChangePin,
+	/// See [`der::Error`].
+	DerError(Box<der::Error>),
+}
+
+impl From<der::Error> for YubiKeyError {
+	fn from(from: der::Error) -> Self {
+		Self::DerError(from.into())
+	}
 }
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs:88]: Makes usage of the already existing `der::Error` enum instead of re-creating the same error variants. The structure is wrapped inside a `Box` to reduce the usage of stack memory.

File: src/qos_client/src/yubikey.rs [hunk 5]
```diff
@@ -91,24 +101,22 @@ pub fn generate_signed_certificate(
...
 	// Create a random serial number
 	let mut serial = [0u8; 20];
 	OsRng.fill_bytes(&mut serial);
 
-	// Don't add any extensions
-	let extensions: &[x509::Extension<'_, &[u64]>] = &[];
-
 	yubikey.verify_pin(pin).map_err(YubiKeyError::FailedToVerifyPin)?;
-	Certificate::generate_self_signed(
+	Certificate::generate_self_signed::<_, p256::NistP256>(
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs:125]: Using `p256::NistP256` because `ALGO` is `EccP256`.

This is also enforced in the implementation. See https://docs.rs/yubikey/latest/yubikey/certificate/yubikey_signer/trait.KeyType.html#impl-KeyType-for-NistP256

File: src/qos_client/src/yubikey.rs [hunk 6]
```diff
@@ -91,24 +101,22 @@ pub fn generate_signed_certificate(
...
 		yubikey,
 		slot,
-		serial,
-		None, // not_after is none so this never expires
-		&[RelativeDistinguishedName::organization("Turnkey")],
+		SerialNumber::new(&serial)?,
+		Validity::from_now(Duration::from_secs(u64::MAX))?,
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs]: The new `yubikey` version demands an expiration time but the current code-base doesn't specific any. `Duration::from_secs(u64::MAX)` is more-or-less maintaining the same behaviour.

File: src/qos_client/src/yubikey.rs [hunk 7]
```diff
@@ -91,24 +101,22 @@ pub fn generate_signed_certificate(
...
+		RdnSequence::from_str("Turnkey")?,
 		public_key_info,
-		extensions,
+		|_| Ok(()),
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs:134]: The new `yubikey` version requires a callback where users can insert extensions instead of a slice.

File: src/qos_client/src/yubikey.rs [hunk 8]
```diff
@@ -135,13 +143,10 @@ pub fn import_key_and_generate_signed_certificate(
...
-	let public_key_info = {
-		let encoded_point = SecretKey::from_be_bytes(key_data)
-			.map_err(|_| YubiKeyError::InvalidSecret)?
-			.public_key()
-			.to_encoded_point(false);
-		PublicKeyInfo::EcP256(encoded_point)
-	};
+	let public_key_info = SecretKey::from_slice(key_data)
+		.ok()
+		.and_then(|sk| SubjectPublicKeyInfo::from_key(sk.public_key()).ok())
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs:166]: It is possible to create `SubjectPublicKeyInfo` from anything that implements `EncodePublicKey`

File: src/qos_client/src/yubikey.rs [hunk 9]
```diff
@@ -314,10 +319,9 @@ pub fn yubikey_piv_reset() -> Result<(), YubiKeyError> {
...
-	public_key_info: &PublicKeyInfo,
+	bytes: Option<&[u8]>,
 ) -> Result<p256::EncodedPoint, YubiKeyError> {
-	match public_key_info {
-		PublicKeyInfo::EcP256(encoded_point) => Ok(*encoded_point),
-		_ => Err(YubiKeyError::FoundNonP256Key),
-	}
+	bytes
+		.and_then(|el| p256::EncodedPoint::from_bytes(el).ok())
+		.ok_or(YubiKeyError::FoundNonP256Key)
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs]: AFAICT `SubjectPublicKeyInfoOwned` structures originated from elliptic curve algorithms already contain encoded points.

Here goes the workflow of `elliptic_curve::sec1::EncodedPoint<NistP256>`:

`elliptic_curve::sec1::EncodedPoint<NistP256>`[^1] ->  `PublicKey::<NistP256>`[^2] -> `pkcs8::SubjectPublicKeyInfo`[^3] -> `der::Document`[^4] -> `SubjectPublicKeyInfoOwned`



[^1]: https://github.com/iqlusioninc/yubikey.rs/blob/7f2b423713f676a131fdc7fe2578b8f73679839c/src/piv.rs#L1190-L1231

[^2]: https://github.com/RustCrypto/traits/blob/d5355ef8a85577d9842505cf3241f6e1d011d0da/elliptic-curve/src/public_key.rs#L409-L411

[^3]: https://github.com/RustCrypto/traits/blob/d5355ef8a85577d9842505cf3241f6e1d011d0da/elliptic-curve/src/public_key.rs#L484-L492

[^4]: https://github.com/RustCrypto/formats/blob/ee30600f303b9206e99af9967414c48290758006/spki/src/spki.rs#L165-L167

File: src/qos_client/src/yubikey.rs (hunk 3 above)
Comment (c410-f3r) [reply src/qos_client/src/yubikey.rs]: The comment has been removed

File: src/qos_client/src/yubikey.rs (hunk 2 above)
Comment (c410-f3r) [reply src/qos_client/src/yubikey.rs]: The comment has been removed

File: src/qos_nsm/src/nitro/mod.rs [hunk 10]
```diff
@@ -254,7 +254,7 @@ fn verify_cose_sign1_sig(
 
 	let pub_key =
 		ee_cert.tbs_certificate.subject_public_key_info.subject_public_key;
-	let key = PublicKey::from_sec1_bytes(pub_key)
+	let key = PublicKey::from_sec1_bytes(pub_key.raw_bytes())
```
Comment (cr-tk) [inline src/qos_nsm/src/nitro/mod.rs]: I think this is subtly wrong, for non-trivial reasons around ASN.1 `BIT STRING` handling and potential decoding edge cases with ignored bits, see https://github.com/tkhq/qos/pull/565#pullrequestreview-3055941536.

This code change is handled by #565 now, so you can drop this change when rebasing on top of it.

Comment (c410-f3r) [reply src/qos_nsm/src/nitro/mod.rs]: Oopss... My bad.

File: src/qos_client/src/yubikey.rs [hunk 11]
```diff
@@ -94,26 +105,24 @@ pub fn generate_signed_certificate(
 	let public_key_info =
 		piv::generate(yubikey, slot, ALGO, PinPolicy::Always, touch_policy)
 			.map_err(|_| YubiKeyError::FailedToGenerateKey)?;
-	let encoded_point = extract_encoded_point(&public_key_info)?;
+	let encoded_point =
+		extract_encoded_point(public_key_info.subject_public_key.as_bytes())?;
 
 	// Create a random serial number
 	let mut serial = [0u8; 20];
```
Comment (cr-tk) [inline src/qos_client/src/yubikey.rs]: It's not a fault of this PR, but our existing code to create a random serial number may be slightly flawed, which now becomes more visible due to this PR.

The yubikey crate's integration test has a similar construction to create a random serial number. Code documentation there suggests that the most significant bit of the 20 byte serial can't be set to 1
https://github.com/iqlusioninc/yubikey.rs/blob/7f2b423713f676a131fdc7fe2578b8f73679839c/tests/integration.rs#L167-L173 . From what I understand, that's because the spec https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.2 requires the serial to be a _positive_ integer, but the data is read as a signed integer, so we should ensure the sign bit isn't flipped to negative.

Based on how we currently ask the OsRng to fill this buffer, there's a 50% chance the most significant bit is set. I expect it'll error out in the new `SerialNumber::new()` usage that this PR introduces when that happens, because the `asn1::Uint` `value_len()` is then 21. [x509-cert includes code](https://github.com/RustCrypto/formats/blob/b579ad6c3a85e117d9f4c025d84f05d51f3eab4a/x509-cert/src/serial_number.rs#L48-L60)  that checks for this, but isn't straightforward to understand.

To fix this, we could 
1. Copy the approach of the yubikey crate to leave the most significant byte at 0x00
2. Mask just the most significant bit. I'm leaning towards that option. 

Since we're using this pattern more than once, and want to have the new serial generation logic testable outside of actual tests with  Yubikey hardware, I'll refactor this a bit and add a test.

Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: Implemented via fcb4fcbeac56bc3c043743be0d0132a5f0cc9990.

**Comment by c410-f3r** [comment]
https://github.com/iqlusioninc/yubikey.rs?tab=readme-ov-file#testing

> To run the full test suite, you'll need a connected YubiKey NEO/4/5 device in the default state (i.e. default PIN/PUK).

NVM, it is not possible to create tests without a physical YubiKey unit. Even the `yubikey` repository ignores their integration tests -> https://github.com/iqlusioninc/yubikey.rs/blob/main/tests/integration.rs

**Comment by c410-f3r** [comment]
I guess I did everything that was within my reach. This PR is now in a ready state, for what it's worth.

File: src/qos_client/src/yubikey.rs (hunk 4 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:88]: @r-n-o : what do you think about this? I'm unsure if the verbosity level of the very specific DER errors (exact parser problem and error position within the message) is necessary for us, and if the `Box<>` wrapping could lead to any difficulties down the road.

My intuition is that we can probably get away with a more high-level generic new `DerProcessingError` or explicitly map the relevant cases into an existing error enum item, like we've been doing in many places.

Specifically, the new error is triggered on 
```
		SerialNumber::new(&serial)?,
		Validity::from_now(Duration::from_secs(u64::MAX))?,
		RdnSequence::from_str("Turnkey")?,
```
when generating a certificate for new/imported keys. Failure results of the overall operations are mapped to `YubiKeyError::FailedToGenerateSelfSignedCert` already, so intermediary failures during its processing could be as well.

File: src/qos_client/src/yubikey.rs (hunk 5 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:125]: Makes sense to me. We're using this for P-256, not P-384.

File: src/qos_client/src/yubikey.rs (hunk 6 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: @c410-f3r unfortunately, I think this is a bug.

`Validity::from_now()` internally _adds_ the `duration` parameter to `SystemTime::now()`. Since it doesn't silently cap at a maximum value but loudly fails in case of overflows, I think this code would have crashed on every iteration, because `u64::MAX + anything positive` is always an overflow :grimacing: 

Example:
```
#[test]
fn test_validity_calculation() {
	let _validity = Validity::from_now(Duration::from_secs(u64::MAX)).unwrap();
}
```

> thread 'yubikey::test_validity_calculation' panicked at library/std/src/time.rs:601:31:
> overflow when adding duration to instant

File: src/qos_client/src/yubikey.rs [hunk 12]
```diff
@@ -29,6 +29,7 @@ pub const SIGNING_SLOT: SlotId = SlotId::Signature;
 /// Factory default pin for yubikeys.
 pub const DEFAULT_PIN: &[u8] = b"123456";
 const ALGO: AlgorithmId = AlgorithmId::EccP256;
+const VALIDITY_SECS: u32 = 60 * 60 * 24 * 365;
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs]: 1 year. Purposefully using an smaller integer because of the `u64::MAX` incident.

File: src/qos_client/src/yubikey.rs [hunk 13]
```diff
@@ -350,21 +351,35 @@ fn generate_random_rfc5280_serial() -> [u8; 20] {
...
+				SerialNumber::new(&serial_data).unwrap();
+		}
+	}
+
+	#[test]
+	#[should_panic]
+	fn test_validity_calculation() {
+		let _validity =
+			Validity::from_now(Duration::from_secs(u64::MAX)).unwrap();
 	}
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs]: Your test has been added for future reference. Feel free to tell if it is not worth including

File: src/qos_client/src/yubikey.rs [hunk 14]
```diff
@@ -350,21 +351,35 @@ fn generate_random_rfc5280_serial() -> [u8; 20] {
 	serial
 }
 
-#[test]
-fn test_rfc5280_serial_generation_success() {
-	use x509_cert::certificate::Rfc5280;
+#[cfg(test)]
+mod tests {
```
Comment (c410-f3r) [inline src/qos_client/src/yubikey.rs:370]: The compiler spends more time eliminating dead code that is not gated behind a `cfg` flag.

File: src/qos_client/src/yubikey.rs (hunk 6 above)
Comment (c410-f3r) [reply src/qos_client/src/yubikey.rs]: Erhhh... It is totally my fault. Hopefully you won't encounter more problems.

Not sure about `SubjectPublicKeyInfoOwned` though 😬

File: src/qos_client/src/yubikey.rs (hunk 13 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: It's generally a good idea to add test coverage after encountering edge cases, but in this particular instance I don't think the test should be added. It was meant as a proof-of-concept to reproduce how this can never work, but we kind of know that - unless Rust changes something major in the background, which I don't expect and wouldn't hurt us, so it doesn't need constant testing IMHO.

File: src/qos_client/src/yubikey.rs (hunk 14 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:370]: Sounds good. It's probably not strictly necessary with one test, but seems like good form.

File: src/qos_client/src/yubikey.rs (hunk 12 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: I'm not confident that 1y is enough. We went from "effectively no expiry" to "some expiry", and 1y can backfire when assumed to be "never", because some certificates may be long-lived and relevant enough for that to bite, which then isn't seen during testing either.

Looking at the original code in `x509` that we previously triggered with the `never` option, there's the following comment and code:
```
    /// From [RFC 5280](https://tools.ietf.org/html/rfc5280#section-4.1):
    /// ```text
    /// Validity ::= SEQUENCE {
    ///      notBefore      Time,
    ///      notAfter       Time }
    ///
    /// To indicate that a certificate has no well-defined expiration date,
    /// the notAfter SHOULD be assigned the GeneralizedTime value of
    /// 99991231235959Z.
    /// ```
    fn validity<W: Write>(
        not_before: DateTime<Utc>,
        not_after: Option<DateTime<Utc>>,
    ) -> impl SerializeFn<W> {
        der_sequence((
            time(not_before),
            time(not_after.unwrap_or_else(|| Utc.ymd(9999, 12, 31).and_hms(23, 59, 59))),
        ))
    }
```
I'm wondering if we should bump the `VALIDITY_SECS` to something like 10y, or restore the previous behavior by setting the special date in year 9999 as outlined in https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5 .

@r-n-o what do you think?

File: src/qos_client/src/yubikey.rs (hunk 7 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:134]: Makes sense to me.

File: src/qos_client/src/yubikey.rs (hunk 8 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:166]: I think that's correct:
```
        /// Create a [`SubjectPublicKeyInfoOwned`] from any object that implements
        /// [`EncodePublicKey`].
```

The `elliptic-curve` crate has such an implementation for `PublicKey`:
```
#[cfg(all(feature = "alloc", feature = "pkcs8"))]
impl<C> EncodePublicKey for PublicKey<C>
```
Both the `alloc` and `pkcs8` feature are turned on in the dependency tree via `yubikey` -> `p256` -> `ecdsa`.

File: src/qos_client/src/yubikey.rs (hunk 6 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: Discussed further https://github.com/tkhq/qos/pull/564#discussion_r2280513964 , thanks for the first patch approach on this.

I think we're okay on `SubjectPublicKeyInfoOwned`, see https://github.com/tkhq/qos/pull/564#discussion_r2282821021 .

File: src/qos_client/src/yubikey.rs [hunk 15]
```diff
@@ -116,7 +117,7 @@ pub fn generate_signed_certificate(
 		yubikey,
 		slot,
 		SerialNumber::new(&serial)?,
-		Validity::from_now(Duration::from_secs(u64::MAX))?,
+		Validity::from_now(Duration::from_secs(VALIDITY_SECS.into()))?,
 		RdnSequence::from_str("Turnkey")?,
```
Comment (cr-tk) [inline src/qos_client/src/yubikey.rs]: @c410-f3r I've run the device-specific integration tests (see [here](https://github.com/tkhq/qos/blob/ae01904c756107f850aea42000137ef124df3fe4/src/qos_client/tests/yubikey.rs#L23-L37) for details) with a spare Yubikey and found another bug:

`RdnSequence::from_str("Turnkey")?,` always fails. It expects a string that conforms to https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.4 , specifically it needs to follow the Distinguished Name format. `Turnkey` fails, but `CN=Turnkey` would work, for example. Same problem for `import_key_and_generate_signed_certificate()`. After changing the strings, the tests run through.

Since QOS can be used by other parties than Turnkey, I think we should switch this identifier to something more generic. I'm leaning towards `"CN=QuorumOS"` without a country (C) or organization (O) field.

@r-n-o : If you have time, a second opinion on the cert identifier would be useful :slightly_smiling_face:

Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: Switched to `"CN=QuorumOS"`. @r-n-o feedback welcome.

File: src/qos_client/src/yubikey.rs (hunk 13 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: Removed in edc7d838fd2d07149873cc0de9e4caa93ca1dae4.

File: src/qos_client/src/yubikey.rs (hunk 12 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: I went with the "easy" solution of 10y in edc7d838fd2d07149873cc0de9e4caa93ca1dae4, feedback welcome.

File: src/qos_client/src/yubikey.rs (hunk 4 above)
Comment (emostov) [reply src/qos_client/src/yubikey.rs:88]: In practice I don't think this would matter either way, but to keep things dead simple and minimize use of 3rd party libs in our public APIs I would propose simply `DerError(String)` and then we can update the `From<der::Error>` impl to just call the display method on the error

The errors here are mostly just for the application level user to understand what are going on and in practice not programmatically processed to make decisions - so I think just know the DerError comes from this lib and the string is the Display of the underlying error should be sufficient and ergonomic

File: src/qos_client/Cargo.toml [hunk 16]
```diff
@@ -15,15 +15,16 @@ qos_nsm = { path = "../qos_nsm", default-features = false }
...
 rand_core = { version = "0.9", features = ["os_rng"], default-features = false }
 zeroize = { version = "1.6", default-features = false }
 rpassword = { version = "7", default-features = false }
 serde_json = { version = "1" }
 
+der = { version = "0.7", default-features = false, optional = true }
 x509 = { version = "0.2", default-features = false, optional = true }
-# As of 7/2025, upgrading this to 0.8.0 is blocked due to dependency conflicts
-yubikey = { version = "0.7", features = ["untested"], default-features = false, optional = true }
+x509-cert = { version = "0.2", default-features = false, optional = true }
```
Comment (emostov) [inline src/qos_client/Cargo.toml]: we can do this in a separate PR, but now that we are using `x509-cert`, can we get away with removing `x509` crate? Seems like unnecessary dependency review burden if we can get away with only one of these

File: src/qos_client/src/yubikey.rs [hunk 17]
```diff
@@ -94,26 +108,21 @@ pub fn generate_signed_certificate(
...
-	Certificate::generate_self_signed(
+	Certificate::generate_self_signed::<_, p256::NistP256>(
 		yubikey,
 		slot,
-		serial,
-		None, // not_after is none so this never expires
-		&[RelativeDistinguishedName::organization("Turnkey")],
+		SerialNumber::new(&serial)?,
+		Validity::from_now(Duration::from_secs(CERTIFICATE_VALIDITY_SECS.into()))?,
+		RdnSequence::from_str("CN=QuorumOS")?,
```
Comment (emostov) [inline src/qos_client/src/yubikey.rs]: nit: can we make `"CN=QuorumOS"` a const

File: src/qos_client/Cargo.toml (hunk 16 above)
Comment (cr-tk) [reply src/qos_client/Cargo.toml]: Well spotted! It seems we don't need `x509` as a direct crate anymore. I just removed it for testing, and the basic tests run through without issue. We also don't have code using any `x509::*` code.

I'll make a quick commit for this.

Comment (cr-tk) [reply src/qos_client/Cargo.toml]: Note that the dependency review for this PR is already done, so it wouldn't have mattered there this time - but every dependency we can remove from the tree will need less maintenance and review in the future :+1:

File: src/qos_client/src/yubikey.rs (hunk 17 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: Done in 09304ba9f9d01157aba36bab6e01721e319ca523.

File: src/qos_client/Cargo.toml (hunk 16 above)
Comment (cr-tk) [reply src/qos_client/Cargo.toml]: Done via 09304ba9f9d01157aba36bab6e01721e319ca523.

**Review by emostov — COMMENTED** [review]
Generally this looks right to me. My only nit would be nesting the error as string details: https://github.com/tkhq/qos/pull/564/files#r2285766408

File: src/qos_client/src/yubikey.rs (hunk 4 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:88]: @emostov addressed in 738101eaa42d91ec139322a960312aa137d24f2e.

File: src/qos_client/src/yubikey.rs [hunk 18]
```diff
@@ -94,26 +111,23 @@ pub fn generate_signed_certificate(
...
 		yubikey,
 		slot,
-		serial,
-		None, // not_after is none so this never expires
-		&[RelativeDistinguishedName::organization("Turnkey")],
+		SerialNumber::new(&serial)?,
+		Validity::from_now(Duration::from_secs(
+			CERTIFICATE_VALIDITY_SECS.into(),
+		))?,
+		RdnSequence::from_str(CERTIFICATE_DISTINGUISHED_NAME)?,
```
Comment (r-n-o) [inline src/qos_client/src/yubikey.rs:132]: This is a behavioral change ("O=Turnkey" => "CN=QuorumOS") but this is a good change I believe.

Just need to be mindful of already-provisioned yubikeys and tooling which may expect "O=Turnkey".

File: src/qos_client/src/yubikey.rs [hunk 19]
```diff
@@ -94,26 +111,23 @@ pub fn generate_signed_certificate(
...
 		public_key_info,
-		extensions,
+		|_| Ok(()),
```
Comment (r-n-o) [inline src/qos_client/src/yubikey.rs:134]: From the docs on this function:
```rust
    /// `extensions` is optional; if empty, no extensions will be included. Due to the
    /// need for an `O: Oid` type parameter, users who do not have any extensions should
    /// use the workaround `let extensions: &[x509_cert::Extension<'_, &[u64]>] = &[];`.
```

which seems incorrect given the argument is a closure now.

Submitted a PR to fix that: https://github.com/iqlusioninc/yubikey.rs/pull/632

File: src/qos_client/src/yubikey.rs [hunk 20]
```diff
@@ -162,24 +173,20 @@ pub fn import_key_and_generate_signed_certificate(
...
-	// Create a random serial number
-	let mut serial = [0u8; 20];
-	OsRng.try_fill_bytes(&mut serial).expect(
-		"The OsRng was unable to provide data, which should never happen",
-	);
-
-	// Don't add any extensions
-	let extensions: &[x509::Extension<'_, &[u64]>] = &[];
+	// Create a random serial number compliant with RFC5280
+	let serial = generate_random_rfc5280_serial();
```
Comment (r-n-o) [inline src/qos_client/src/yubikey.rs:181]: Nice!

File: src/qos_client/src/yubikey.rs [hunk 21]
```diff
@@ -324,10 +334,54 @@ pub fn yubikey_piv_reset() -> Result<(), YubiKeyError> {
 }
 
 fn extract_encoded_point(
-	public_key_info: &PublicKeyInfo,
+	bytes: Option<&[u8]>,
 ) -> Result<p256::EncodedPoint, YubiKeyError> {
```
Comment (r-n-o) [inline src/qos_client/src/yubikey.rs:340]: This interface is a bit awkward: I'd pick `&[u8]` or `BitStringRef` as an argument type. The fact that `as_bytes` returns an `Option` is not something that should be known to a generic helper operating on bytes. My proposal would be to make it it as easy on callers as possible:

```rust
fn extract_encoded_point(
	public_key_info: BitStringRef,
) -> Result<p256::EncodedPoint, YubiKeyError> {
	public_key_info.as_bytes()
		.ok_or(YubiKeyError::EmptyPublicKeyInfo)
		.and_then(|el| p256::EncodedPoint::from_bytes(el).ok())
		.ok_or(YubiKeyError::FoundNonP256Key)
}
```

And callers can then use this with `extract_encoded_point(public_key_info.subject_public_key)` or `extract_encoded_point(public_key_info.subject_public_key.owned_to_ref())?;`

File: src/qos_client/src/yubikey.rs [hunk 22]
```diff
@@ -324,10 +334,54 @@ pub fn yubikey_piv_reset() -> Result<(), YubiKeyError> {
...
-	match public_key_info {
-		PublicKeyInfo::EcP256(encoded_point) => Ok(*encoded_point),
-		_ => Err(YubiKeyError::FoundNonP256Key),
+	bytes
+		.and_then(|el| p256::EncodedPoint::from_bytes(el).ok())
```
Comment (r-n-o) [inline src/qos_client/src/yubikey.rs]: If `bytes` is `None` we should have an error that is specific to this case, something like `YubiKeyError::NoPublicKeyInfoBytes`. Otherwise we'll trigger `YubiKeyError::FoundNonP256Key` when `bytes` is `None`, which is confusing: not having bytes isn't the same as having non-P256 bytes!

File: src/qos_client/src/yubikey.rs (hunk 18 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:132]: Yes, see https://github.com/tkhq/qos/pull/564#discussion_r2282940043 for more thoughts here. We should test this carefully when making the change.

File: src/qos_client/src/yubikey.rs (hunk 19 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:134]: Good catch! We're good on our end with the current construction, correct?

Comment (r-n-o) [reply src/qos_client/src/yubikey.rs:134]: Yep! This comment was just a note; nothing to fix!

File: src/qos_client/src/yubikey.rs (hunk 22 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs]: Good point. Understood broadly, a "this is not a p256 key" error would be true if there is no pubkey data, but our existing wording of `YubiKeyError::FoundNonP256Key` suggests there _was_ a key, just not a _p256_ one.

New error added in 9e3bd8fc51777178b6d3a746a0a3c0263bde4954.

File: src/qos_client/src/yubikey.rs (hunk 21 above)
Comment (cr-tk) [reply src/qos_client/src/yubikey.rs:340]: Well spotted, the `Options` interface is awkward.

I've followed your suggestion in 9e3bd8fc51777178b6d3a746a0a3c0263bde4954 , but had to bend it into shape a bit. @r-n-o let me know if the implementation fits your expectations.
